### PR TITLE
refactor(core): add support for `on timer` trigger in `@defer` blocks

### PR DESCRIPTION
**Pull request by AndrewKushnir**
This commit adds the logic to support `on timer` triggers in `@defer` blocks in both rendering and prefetching conditions.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/test/acceptance/defer_spec.ts [hunk 1]
```diff
@@ -57,6 +57,15 @@ function allPendingDynamicImports() {
   return dynamicImportOf(null, 10);
 }
 
+/**
+ * Invoke a callback function after a specified amount of time (in ms).
+ */
+function timer(delay: number): Promise<void> {
+  return new Promise<void>((resolve) => {
```
Comment (crisbeto) [inline packages/core/test/acceptance/defer_spec.ts:64]: nit: you can save one function instance here like this: `new Promise<void>(resolve => setTimeout(resolve, delay));`

File: packages/core/src/render3/instructions/defer.ts [hunk 2]
```diff
@@ -1056,3 +1109,173 @@ class OnIdleScheduler {
     factory: () => new OnIdleScheduler(),
   });
 }
+
+/**
+ * Helper service to schedule `setTimeout`s for batches of defer blocks,
+ * to avoid calling `setTimeout` for each defer block (e.g. if defer blocks
+ * are created inside a for loop).
+ */
+class TimerScheduler {
```
Comment (crisbeto) [inline packages/core/src/render3/instructions/defer.ts:1118]: Should we move it out into a separate file? We ended up accumulating a lot in the defer instructions file.

Comment (jessicajaniuk) [reply packages/core/src/render3/instructions/defer.ts:1118]: +1

File: packages/core/src/render3/instructions/defer.ts [hunk 3]
```diff
@@ -1056,3 +1109,173 @@ class OnIdleScheduler {
...
+      }
+    }
+    arrayInsert2(target, insertAtIndex, invokeAt, callback);
+  }
+
+  private removeFromQueue(target: Array<number|VoidFunction>, callback: VoidFunction) {
+    let index = -1;
+    for (let i = 0; i < target.length; i += 2) {
+      const queuedCallback = target[i + 1];
+      if (queuedCallback == callback) {
```
Comment (jessicajaniuk) [inline packages/core/src/render3/instructions/defer.ts:1176]: Is this `==` intentional vs a `===`?

**Review by jessicajaniuk — APPROVED** [review]
LGTM aside from the non-blocking comments.

File: packages/core/src/render3/instructions/defer.ts (hunk 3 above)
Comment (AndrewKushnir) [reply packages/core/src/render3/instructions/defer.ts:1176]: Thanks for catching this! 👍

File: packages/core/src/render3/instructions/defer.ts (hunk 2 above)
Comment (AndrewKushnir) [reply packages/core/src/render3/instructions/defer.ts:1122]: Yes, I want to create a `defer` folder and split the `defer.ts` file into multiple ones and put them into the `defer` folder. I'll take a note on this and will do the refactoring after the feature freeze.

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:570191343:BASE:570191381:1696284343581:e8dd7fa4).

File: packages/core/test/acceptance/defer_spec.ts (hunk 1 above)
Comment (AndrewKushnir) [reply packages/core/test/acceptance/defer_spec.ts:64]: Thanks for the comment. Will do a cleanup in a followup PR (there are some existing functions with the same pattern above, which would require a similar update).

**Comment by AndrewKushnir** [comment]
**Caretaker note**: presubmit is "green" (only unrelated failures), this PR is ready for merge.

**Review by crisbeto — APPROVED** [review]

**Comment by alxhub** [comment]
This PR was merged into the repository by commit 650ab4f21809896a2b576ac8a35a63733a53497f.

**Pull request closed without merging**